### DISCUSSION TITLE
Adding offset to IfcBeam's profile

### DISCUSSION THREAD

**Discussion by zooid** [IfcOpenShell/IfcOpenShell#4939] [Q&A] [answered]
Where is this offset applied?
is it the profile, the IfcBeam or is it in the IfcExtrudedAreaSolid (that one seems to be the "puppet master")

https://standards.buildingsmart.org/IFC/RELEASE/IFC2x3/TC1/HTML/ifcprofileresource/lexical/ifcrectanglehollowprofiledef.htm
says only that:
```
Position
The parameterized profile defines its own position coordinate system. The underlying coordinate system is defined by the swept area solid that uses the profile definition. It is the xy plane of:

IfcSweptAreaSolid.Position
by using offsets of the position location, the parameterized profile can be positioned centric (using x,y offsets = 0.), or at any position relative to the profile. Explicit coordinate offsets are used to define cardinal points (e.g. upper-left bound).
```

But without information what it is that needs to be done.

**Comment by Moult** [comment]
The profile defines your cross sectional profile in 2D.

The extrusion selects the profile, as well as defines the extrusion depth and angle (typically 0,0,1) and the "position" of the profile. I.e. is the profile extruded by its centerpoint, its top left, its bottom middle, etc. This is known semantically as the "cardinal point".

The beam simply contains the extrusion and positions it in the world via its placement matrix (location and rotation, as extrusions only go +Z).

Here is an example showing a lot of variations you can see:

[test.ifc.txt](https://github.com/user-attachments/files/16022306/test.ifc.txt)

I advise using the BlenderBIM Add-on which shows these placement points very clearly. Notice how the orange dot is different for each beam.

![2024-06-28-094623_830x425_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/be7fb045-c189-49ac-9d6c-b03d7157003e)

**Comment by zooid** [comment]
I've tried creating yet another "axis system" or "plane" I'm not really sure what it actually is `IfcAxis2Placement3D`

```python
profile_plane = _model.create_entity("IfcAxis2Placement3D",
                                    Location=position_point,
                                    Axis=extrusion_direction,
                                    RefDirection=local_x_axis)

    offsets_to_profile = _model.create_entity("IfcAxis2Placement3D",
                                              Location=position_point,
                                              Axis=_model.create_entity("IfcDirection", DirectionRatios=(0., 0., 1.)),
                                              RefDirection=_model.create_entity("IfcDirection", (0.0, 0.0, float(_profile_half_height))))
    if _is_column:
        extruded_area_solid = _model.create_entity("IfcExtrudedAreaSolid",
                                                   SweptArea=_profile,
                                                   Position=profile_plane,
                                                   ExtrudedDirection=_model.create_entity("IfcDirection", (0.0, 0.0, 1.0)),
                                                   Depth=_extrusion_length)

    else:
        extruded_area_solid = _model.create_entity("IfcExtrudedAreaSolid",
                                                   SweptArea=_profile,
                                                   Position=offsets_to_profile,
                                                   ExtrudedDirection=_model.create_entity("IfcDirection", (0.0, 0.0, 1.0)),
                                                   Depth=_extrusion_length)
```
but messing with it only rotates my profile or make it become just a line, instead of moving it up (on top of the columns)

Original:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/dda58da8-b179-4ee1-b26c-834354964535)

attempt using Axis=extrusion_direction:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/3a981a23-2e21-41a1-bd93-e21745717875)

attempt setting Axis to global X axis:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/7cdfc5fc-df8d-4bba-99a3-a4eb4ab12fca)

**Reply by zooid** [reply, marked as answer]
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/10072821-498e-4768-a740-ebc2d194d8d3)

Jackpot!
